File: basicCalculatorII-LC 227.py
# Time Complexity = O(n)
# Space Complexity = O(1)

class Solution:
    def calculate(self, s: str) -> int:
        if len(s) == 0:
            return 0
        
        s = s.strip()       # To trim the extra white spaces
        s = s.replace(" ", "")
        
        
        last_sign = '+'
        calc = 0
        num = 0
        tail = 0
        
        for i in range(len(s)):
            char = s[i]
            if char.isdigit():
                num = num * 10 + ord(char) - ord('0')
                
                
            if not char.isdigit() and char != '' or i == len(s) - 1:
                if last_sign == '+':
                    calc = calc + num
                    tail = + num
                
                elif last_sign == '-':
                    calc = calc - num
                    tail = - num
                
                elif last_sign == '*':
                    calc = calc - tail + (tail * num)
                    tail = tail * num
                
                else:
                    calc = calc - tail + int(float(tail / num))
                    tail = int(float(tail / num))
                    
                    
                last_sign = char
                num = 0
                
        return calc
   
    
    
# Stack Solution 
# Time Complexity = O(n)
# Space Complexity = O(n), recursive stack

class Solution:
    def calculate(self, s: str) -> int:
        if len(s) == 0:
            return 0
        
        s = s.strip()       # To trim the extra white spaces
        s = s.replace(" ", "")
        
        
        last_sign = '+'
        calc = 0
        num = 0
        stack = []
        
        for i in range(len(s)):
            char = s[i]
            if char.isdigit():
                num = num * 10 + ord(char) - ord('0')
                
                
            if not char.isdigit() and char != '' or i == len(s) - 1:
                if last_sign == '+':
                    stack.append(num)
                
                elif last_sign == '-':
                    stack.append(- num)
                
                elif last_sign == '*':
                    stack.append(stack.pop() * num)
                
                else:
                    stack.append(int(float(stack.pop() / num)))
                    
                    
                last_sign = char
                num = 0
        
        while stack:
            calc += stack.pop()
                
        return calc

File: BasicCalculatorIILC227.java
class Solution {
    
   
    //Time Complexity: O(n), n = length of s
    //Space Complexity: O(1)
    
    public int calculate(String s) {
     
        int result = 0;
        int currNum = 0;
        int tail = 0;
        char lastSign = '+';
        
        for(int i=0; i<s.length(); i++){
            
            char c = s.charAt(i);
            
            if(Character.isDigit(c)){
                
                currNum = currNum * 10 + c - '0'; 
            }
            
            if((!Character.isDigit(c) && c != ' ') || i == s.length() - 1){
                
                if (lastSign == '+'){
                    
                    result = result + currNum;
                    tail = currNum;
                }
                else if (lastSign == '-'){
                    
                    result = result - currNum;
                    tail = -currNum;
                }
                else if (lastSign == '*'){
                    
                    result = result - tail + tail * currNum;
                    tail = tail * currNum;
                }
                else if (lastSign == '/'){
                    
                    result = result - tail + tail / currNum;
                    tail = tail / currNum;
                }
                
                currNum = 0;
                lastSign = c;
            }   
        }
        return result;
    }
    
    

    
    
    
    //Time Complexity: O(n), n = length of s
    //Space Complexity: O(n)
    
//     public int calculate(String s) {
        
//         Stack<Integer> stack = new Stack<>();
//         int currNum = 0;
//         char lastSign = '+';
        
//         for(int i=0; i<s.length(); i++){
          
//             char c = s.charAt(i);
            
//             if(Character.isDigit(c)){
                
//                 currNum = currNum * 10 + c -'0';
                
//             }
//             if((!Character.isDigit(c) && c != ' ') || i == s.length() - 1){
                
//                 if (lastSign == '+') stack.push(currNum);
//                 else if (lastSign == '-') stack.push(-currNum);
//                 else if (lastSign == '*') stack.push(stack.pop() * currNum);
//                 else if (lastSign == '/') stack.push(stack.pop() / currNum);
                
//                 currNum = 0;
//                 lastSign = c;
//             }
//         }
        
//         int result = 0;
//         while(!stack.isEmpty()){
//             result += stack.pop();
//         }
//         return result;
//     }
}
